### DISCUSSION TITLE
Worlds

### DISCUSSION THREAD

**Discussion by fredchristianson** [fredchristianson/javascript-game-engine#8] [Ideas]
I am calling the main game area the _**World**_.  We need to consider different types of worlds we may want in order to architect the classes and functions to support them:
- Simple 2D board (tic-tac-to, monopoly, cards, flat dice, flat tiles)
- 3D board (dice cubes, box tiles)
- Perspective board (nearer dice cubes are larger)
- Image (background, side-scroller landscape,  scrollable horiz/vert map)
- Drawable (canvas or SVG with user or javascript drawing)
- Multi-layered (2 or more of the others)

We want to come up with a class hierarchy and operations to implement as much as possible in the core.  It will be something along these lines.
![image](https://user-images.githubusercontent.com/3279104/223867413-bc899824-a2dc-4898-9f75-f9c1ced3a09e.png)

There will likely be another hierarchy that does the "drawing" in different ways
- HTML/CSS
- Canvas
- SVG
Most layer types can be drawing with any of the drawing types.
Determining what operations are available in different classes is a key architectural decisions.  
We also need to determining the public operations and the implementation operations.  JavaScript doesn't have "protected".  We can implement protected members by naming conventions (e.g. this._protectedMethod()) or by separating public classes from implementation classes.

**Comment by Eddie7145** [comment]
So basically I'm this project we will have different  "Worlds" and each world so to speak has its multiple layer that make up that world

**Reply by fredchristianson** [reply]
I will try to use the term GameArea instead of "World".  

The project will have 1 app with multiple games.

Each Game will have one GameArea

Each GameArea will have 1 or more layers.
